### ISSUE TITLE
MIME type parsing

### ISSUE THREAD

**Issue by annevk** [httpwg/http-core#33]
See http://www.rfc-editor.org/errata_search.php?rfc=7231&eid=4031 (rejected due to scope). Clients typically (or all?) allow `text/html;` despite it being invalid.

**Comment by annevk** [comment]
One of the things that I couldn't find addressed anywhere is something like `text/html;charset=gbk;charset=utf-8`. The introduction of https://tools.ietf.org/html/rfc2046 mentions parameters as a set, thereby potentially outlawing duplicate names, but there's nothing normative to that effect anywhere. (Browsers are evenly split on how they handle this case by the way.)

**Comment by mnot** [comment]
See also https://github.com/whatwg/mimesniff/issues/30 e.g., discussion of multiple content-types. This has potential security impact.

**Comment by annevk** [comment]
For MIME types I ended up (re)writing https://mimesniff.spec.whatwg.org/#understanding-mime-types btw and also have a bunch of tests at https://github.com/w3c/web-platform-tests/tree/master/mimesniff/mime-types.

**Comment by mnot** [comment]
Discussed in Basel; inclination is to change the ABNF, and survey similar places to align if necessary.

**Comment by snuggs** [comment]
@mnot pardon my ignorance but what is ABNF? Am I wrong to assume [it is this](https://en.wikipedia.org/wiki/Augmented_Backus%E2%80%93Naur_form)

Thanks in advance!

**Comment by mnot** [comment]
It is.

**Comment by mnot** [comment]
Note that this is about the trailing semicolon, at least initially.

**Comment by mnot** [comment]
Discussed on editors' call; we should create a new ABNF rure `parameters` in 5.4.1.4. Parameters that allows semicolons without following `parameter`, and use that where appropriate.

**Comment by reschke** [comment]
Just to check (grep in httpbis.abnf):
```
expectation = token [ "=" ( token / quoted-string ) *parameter ]

media-range = ( "*/*" / ( type "/*" ) / ( type "/" subtype ) ) *( OWS ";" OWS parameter )
media-type = type "/" subtype *( OWS ";" OWS parameter )

parameter = parameter-name "=" parameter-value
parameter-name = token
parameter-value = ( token / quoted-string )

transfer-coding = token *( OWS ";" OWS transfer-parameter )
transfer-parameter = token BWS "=" BWS ( token / quoted-string )
```

So "expectation" needs a fix (see #420). So "Expect" would be the only affected field in addition to "Content-Type". (Because "TE" is special). Are we ok with that?

That said, do we want:

```
parameters = *( OWS ";" OWS [ parameter ] )
```

or 

```
parameters = *( OWS ";" OWS parameter ) [ OWS ";" ]
```

?

**Comment by reschke** [comment]
See https://github.com/httpwg/http-core/pull/427#issuecomment-669017319
